sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -18,91 +18,79 @@
-      RecordFilter filter) throws
-      TransientKinesisException {
-
+      RecordFilter filter) throws TransientKinesisException {
     this.checkpoint = checkNotNull(initialCheckpoint, "initialCheckpoint");
     this.filter = checkNotNull(filter, "filter");
     this.kinesis = checkNotNull(simplifiedKinesisClient, "simplifiedKinesisClient");
     shardIterator = checkpoint.getShardIterator(kinesis);
   }
 
-  /**
-   * Returns record if there's any present.
-   * Returns absent() if there are no new records at this time in the shard.
-   */
-  public CustomOptional<KinesisRecord> next() throws TransientKinesisException {
-    readMoreIfNecessary();
-
-    if (data.isEmpty()) {
-      return CustomOptional.absent();
-    } else {
-      KinesisRecord record = data.removeFirst();
-      checkpoint = checkpoint.moveAfter(record);
-      return CustomOptional.of(record);
-    }
+  List<KinesisRecord> readNextBatch() throws TransientKinesisException {
+    GetKinesisRecordsResult response = fetchRecords();
+    LOG.debug("Fetched {} new records", response.getRecords().size());
+
+    List<KinesisRecord> filteredRecords = filter.apply(response.getRecords(), checkpoint);
+    millisBehindLatest = response.getMillisBehindLatest();
+    return filteredRecords;
   }
 
-  private void readMoreIfNecessary() throws TransientKinesisException {
-    if (data.isEmpty()) {
-      GetKinesisRecordsResult response = fetchRecords();
-      data.addAll(filter.apply(response.getRecords(), checkpoint));
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    try {
+      GetKinesisRecordsResult response = kinesis.getRecords(shardIterator,
+          checkpoint.getStreamName(), checkpoint.getShardId());
+      shardIterator = response.getNextShardIterator();
+      return response;
+    } catch (ExpiredIteratorException e) {
+      LOG.info("Refreshing expired iterator", e);
+      shardIterator = checkpoint.getShardIterator(kinesis);
+      return fetchRecords();
     }
   }
 
-  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
-    GetKinesisRecordsResult response = null;
-    do {
-      try {
-        response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
-            checkpoint.getShardId());
-        shardIterator = response.getNextShardIterator();
-      } catch (ExpiredIteratorException e) {
-        LOG.info("Refreshing expired iterator", e);
-        shardIterator = checkpoint.getShardIterator(kinesis);
-      }
-    } while (response == null || gotEmptyResponseButIsBeforeEndOfTheStream(response));
-    return response;
+  ShardCheckpoint getCheckpoint() {
+    return checkpoint;
   }
 
-  private boolean gotEmptyResponseButIsBeforeEndOfTheStream(GetKinesisRecordsResult response) {
-    return response.getRecords().isEmpty() && response.getMillisBehindLatest() > 0;
+  boolean isUpToDate() {
+    return millisBehindLatest == 0L;
```

Comment:
As far as I remember in one of the previous reviews you've said, that Kinesis may say millisBehindLatest even if there are actually some records left in the backlog, is that right?
In this case you'll know say that shard is up to date and thus advance the watermark in KinesisReader to Instant.now(). Shouldn't we actually wait some time before we do it?

Reply:
Kinesis may return 0 records with some non-zero millisBehindLatest value while there are still records left further in the shard. So it won't affect currently proposed logic for watermark advancing.